File: lib/board.erl
-module(board).
-behavior(gen_server).

%% board API functions
-export([start_link/0, stop/1, get_all/0, get/1, add/3, remove/2]).
% TODO also create add/2 and remove/1?

%% gen_server callback functions
-export([init/1, handle_call/3, handle_cast/2, 
         handle_info/2, terminate/2, code_change/3]).

-define(SERVER, ?MODULE).

-record(state, {up = no_board, down = no_board, 
                right = no_board, left = no_board}).

%% Board API
%% Due to the nature of this process, all requests are handled synchronously.

%% Starts a board process.
start_link() ->
    RegisteredName = {local, ?SERVER},
    Args = ok,
    Options = [],
    gen_server:start_link(RegisteredName, ?SERVER, Args, Options).

%% Stops a board process.
stop(Board) -> gen_server:call(Board, stop).

%% Gives back a list of all boards ([{Board, node()}].
get_all() -> [{?SERVER, Node} || Node <- [node() | nodes()]].

% Gives back the pid of an adjacent board (or no_board).
get(left) ->  gen_server:call(?SERVER, {get, left});
get(right) -> gen_server:call(?SERVER, {get, right});
get(up) ->    gen_server:call(?SERVER, {get, up});
get(down) ->  gen_Server:call(?SERVER, {get, down}).

%% BoardX = {?SERVER, node()}.
add(Board1, left_of, Board2) ->
    ok = gen_server:call(Board2, {add, Board1, left}),
    ok = gen_server:call(Board1, {add, Board2, right});
add(Board1, down_of, Board2) ->
    ok = gen_server:call(Board2, {add, Board1, down}),
    ok = gen_server:call(Board1, {add, Board2, up});
add(Board1, right_of, Board2) ->
    add(Board2, left_of, Board1);
add(Board1, up_of, Board2) ->
    add(Board2, down_of, Board1).

%% Removes a board
remove(_, no_board) ->     ok;
remove(left_of, Board) ->  ok = gen_server:call(Board, {remove, left});
remove(right_of, Board) -> ok = gen_server:call(Board, {remove, right});
remove(up_of, Board) ->    ok = gen_server:call(Board, {remove, up});
remove(down_of, Board) ->  ok = gen_server:call(Board, {remove, down}).

%% gen_server callbacks:

init(ok) ->
    process_flag(trap_exit, true),
    BeginState = #state{},
    {ok, BeginState}.

handle_call({get, left}, _From, State = #state{left = Left}) ->
    {reply, Left, State};
handle_call({get, right}, _From, State = #state{right = Right}) ->
    {reply, Right, State};
handle_call({get, up}, _From, State = #state{up = Up}) ->
    {reply, Up, State};
handle_call({get, down}, _From, State = #state{down = Down}) ->
    {reply, Down, State};

handle_call({add, OtherBoard, left}, _From, State = #state{left = no_board}) ->
    {reply, ok, State#state{left = OtherBoard}};
handle_call({add, OtherBoard, right}, _From, State = #state{right = no_board}) ->
    {reply, ok, State#state{right = OtherBoard}};
handle_call({add, OtherBoard, up}, _From, State = #state{up = no_board}) ->
    {reply, ok, State#state{up = OtherBoard}};
handle_call({add, OtherBoard, down}, _From, State = #state{down = no_board}) ->
    {reply, ok, State#state{down = OtherBoard}};

handle_call({remove, left}, _From, State = #state{}) ->
    {reply, ok, State#state{left = no_board}};
handle_call({remove, right}, _From, State = #state{}) ->
    {reply, ok, State#state{right = no_board}};
handle_call({remove, up}, _From, State = #state{}) ->
    {reply, ok, State#state{up = no_board}};
handle_call({remove, down}, _From, State = #state{}) ->
    {reply, ok, State#state{down = no_board}};

handle_call(stop, _From, State = #state{}) ->
    Reply = ok,
    {reply, normal, Reply, State}.

handle_cast(_Request, State) ->
    {noreply, State}.

handle_info(_Request, State) ->
    {noreply, State}.

terminate(_Reason, _State = #state{left = Left, right = Right, 
                                    up = Up, down = Down}) ->
    remove(left_of, Right),
    remove(right_of, Left),
    remove(down_of, Up),
    remove(up_of, Down),
    ok.

code_change(_OldVersion, State, _Extra) ->
    {ok, State}.

File: lib/board.ex
defmodule Snake.Board do
  alias Snake.Board
  alias Snake.BoardManager

  @moduledoc """
  Module containing logic to link multiple playing boards together.
  """

  @server __MODULE__

  defmodule State do
    defstruct up: :no_board, down: :no_board,
              right: :no_board, left: :no_board
  end

  # Board API
  # Due to the nature of this process, all requests are handled synchronously.

  @doc """
  Starts a board process.
  """
  def start_link, do: GenServer.start_link(@server, :ok, [name: @server])

  @doc """
  Stops a board process.
  """
  def stop, do: :ok = @server |> GenServer.call :stop

  @doc """
  Returns a tuple with which you can address the board on this node easily.
  """
  def local_board, do: {@server, Node.self}

  @doc """
  Gives back a list of all boards ([{Board, Node.self}]).
  """
  def get_all, do: [Node.self | Node.list] |> Enum.map &({Board, &1})

  @doc """
  Gives back an adjacent board (or :no_board) of the board on this node.
  """
  def get(:left), do: local_board |> GenServer.call {:get, :left}
  def get(:right), do: local_board |> GenServer.call {:get, :right}
  def get(:up), do: local_board |> GenServer.call {:get, :up}
  def get(:down), do: local_board |> GenServer.call {:get, :down}

  @doc """
  Gives back an adjacent board (or :no_board) of the board.
  """
  def get(:left_of, board), do: board |> GenServer.call {:get, :left}
  def get(:right_of, board), do: board |> GenServer.call {:get, :right}
  def get(:up_of, board), do: board |> GenServer.call {:get, :up}
  def get(:down_of, board), do: board |> GenServer.call {:get, :down}

  @doc """
  Adds one board adjacent to another board. (boardX = {@server, node})
  Parameter board can be of the form {Board, node} or a Pid.
  """
  def add(:no_board, _, _board), do: :ok
  def add(_board, _, :no_board), do: :ok
  def add(board1, :left_of, board2) do
    :ok = board2 |> GenServer.call {:add, board1, :left}
    :ok = board1 |> GenServer.call {:add, board2, :right}
  end
  def add(board1, :down_of, board2) do
    :ok = board2 |> GenServer.call {:add, board1, :down}
    :ok = board1 |> GenServer.call {:add, board2, :up}
  end
  def add(board1, :right_of, board2), do: add(board2, :left_of, board1)
  def add(board1, :up_of, board2), do: add(board2, :down_of, board1)

  @doc """
  Removes a board adjacent to the board that is passed in the parameters.
  Parameter board can be of the form {Board, node} or a Pid.
  """
  def remove(_, :no_board), do: :ok
  def remove(:left_of, board) do
    :ok = board |> GenServer.call {:remove, :left}
  end
  def remove(:right_of, board) do
    :ok = board |> GenServer.call {:remove, :right}
  end
  def remove(:up_of, board) do
    :ok = board |> GenServer.call {:remove, :up}
  end
  def remove(:down_of, board) do
    :ok = board |> GenServer.call {:remove, :down}
  end

  # GenServer callbacks

  @doc false
  def init(:ok) do
    Process.flag(:trap_exit, true)
    {:ok, %State{}}
  end

  @doc false
  def handle_call({:get, :left}, _from, state = %State{left: left}) do
    {:reply, left, state}
  end
  def handle_call({:get, :right}, _from, state = %State{right: right}) do
    {:reply, right, state}
  end
  def handle_call({:get, :up}, _from, state = %State{up: up}) do
    {:reply, up, state}
  end
  def handle_call({:get, :down}, _from, state = %State{down: down}) do
    {:reply, down, state}
  end
  
  def handle_call({:add, other_board, :left}, _from,
                  state = %State{left: :no_board}) do
    {:reply, :ok, %State{state | left: other_board}}
  end
  def handle_call({:add, other_board, :right}, _from,
                  state = %State{right: :no_board}) do
    {:reply, :ok, %State{state | right: other_board}}
  end
  def handle_call({:add, other_board, :up}, _from,
                  state = %State{up: :no_board}) do
    {:reply, :ok, %State{state | up: other_board}}
  end
  def handle_call({:add, other_board, :down}, _from,
                  state = %State{down: :no_board}) do
    {:reply, :ok, %State{state | down: other_board}}
  end
  
  def handle_call({:remove, :left}, _from, state = %State{}) do
    {:reply, :ok, %State{state | left: :no_board}}
  end
  def handle_call({:remove, :right}, _from, state = %State{}) do
    {:reply, :ok, %State{state | right: :no_board}}
  end
  def handle_call({:remove, :up}, _from, state = %State{}) do
    {:reply, :ok, %State{state | up: :no_board}}
  end
  def handle_call({:remove, :down}, _from, state = %State{}) do
    {:reply, :ok, %State{state | down: :no_board}}
  end
  
  def handle_call(:stop, _from, state = %State{}) do
    reply = :ok
    {:stop, :normal, reply, state}
  end
  def handle_call(_request, state) do
    {:reply, {:error, :not_supported}, state}
  end

  @doc false
  def terminate(_reason, %State{}) do
    # Notify manager it has to remove this board from all the adjacent boards
    :ok = BoardManager.notify_board_gone(Node.self)
  end
end
